### PR TITLE
Store accounting data in VAST native format

### PR DESCRIPTION
**Pull request by tobim**
This PR will change the ACCOUNTANT to relay its data back to the IMPORTER in the form of table slices. That will allow us to query the data flexibly and export it in the desired format.

TODO:
- [x] Have a table_slice_builder in ACCOUNTANT
- [x] Periodically send slices to IMPORTER
- [x] Add integration tests

### CODE REVIEW

File: libvast/src/system/accountant.cpp [hunk 1]
```diff
@@ -179,9 +241,8 @@ accountant_type::behavior_type accountant(accountant_actor* self,
...
             time ts = std::chrono::system_clock::now();
             for (const auto& [key, value] : r) {
-              auto f = [&, key = key](const auto& x) {
-                record(self, key, x, ts);
-              };
+              auto f
+                = [&, key = key](const auto& x) { record(self, key, x, ts); };
               caf::visit(f, value);
             }
           },
```
Comment (mavam) [inline libvast/src/system/accountant.cpp]: This lambda sequence looks like it could benefit from re-formatting. However, you did the right thing by not doing it together with this commit, which makes the changes stand out clearly.

File: libvast/src/system/accountant.cpp [hunk 2]
```diff
@@ -134,21 +175,42 @@ accountant_type::behavior_type accountant(accountant_actor* self,
...
+      VAST_TRACE(self, "was asked for", num, "slices and produced", produced,
+                 ".", st.slice_buffer.size(), "are remaining in buffer");
+    },
+    // done?
+    [](const bool&) { return false; });
   return {[=](announce_atom, const std::string& name) {
             self->state.actor_map[self->current_sender()->id()] = name;
             self->monitor(self->current_sender());
+            if (name == "importer") {
+              VAST_DEBUG(self, "importer is announced");
```
Comment (mavam) [inline libvast/src/system/accountant.cpp]: Just for clarification: does this message still result in our *subject-verb-object* style of logging?

File: libvast/src/system/accountant.cpp [hunk 3]
```diff
@@ -134,21 +175,42 @@ accountant_type::behavior_type accountant(accountant_actor* self,
...
+      auto& st = self->state;
+      size_t produced = 0;
+      while (num-- > 0 && !st.slice_buffer.empty()) {
+        auto& slice = st.slice_buffer.front();
+        produced += slice->rows();
+        out.push(std::move(slice));
+        st.slice_buffer.pop();
+      }
+      VAST_TRACE(self, "was asked for", num, "slices and produced", produced,
+                 ".", st.slice_buffer.size(), "are remaining in buffer");
```
Comment (mavam) [inline libvast/src/system/accountant.cpp]:
```suggestion
                 ";", st.slice_buffer.size(), "are remaining in buffer");
```
Since none of our logging messages and with a period, maybe `;` seems like a more natural separator.

File: libvast/src/system/accountant.cpp [hunk 4]
```diff
@@ -81,6 +99,29 @@ void record(accountant_actor* self, const std::string& key, T x,
...
+                    {"key", string_type{}}, {"value", string_type{}}}
+          .name("vast.account");
+
+    auto slice_type = get_or(sys.config(), "system.table-slice-type",
+                             defaults::system::table_slice_type);
+    st.builder = factory<table_slice_builder>::make(slice_type, layout);
+    VAST_DEBUG(self, "obtained builder");
+  }
+  if (!st.builder->add(ts, to_string(node), actor_id, st.actor_map[actor_id],
+                       key, to_string(x)))
```
Comment (mavam) [inline libvast/src/system/accountant.cpp]: Can this actually fail? If not (because the programmer dictates the schema), then we should probably make this an assertion.

File: libvast/src/system/accountant.cpp [hunk 5]
```diff
@@ -64,12 +68,26 @@ void init(accountant_actor* self, const path& filename) {
...
+                         defaults::system::table_slice_size);
+}
+
+void finish_slice(accountant_actor* self) {
+  auto& st = self->state;
+  // Do nothing if builder has not been created.
+  if (!st.builder)
+    return;
+  auto slice = st.builder->finish();
+  VAST_DEBUG(self, "generated slice with", slice->rows(), "rows");
```
Comment (mavam) [inline libvast/src/system/accountant.cpp]: Maybe add an assertion here that `slice->rows() > 0`?

File: libvast/vast/fwd.hpp [hunk 6]
```diff
@@ -91,6 +91,9 @@ void intrusive_ptr_add_ref(const table_slice*);
 void intrusive_ptr_release(const table_slice*);
 table_slice* intrusive_cow_ptr_unshare(table_slice*&);
 
+void intrusive_ptr_add_ref(const table_slice_builder*);
+void intrusive_ptr_release(const table_slice_builder*);
```
Comment (mavam) [inline libvast/vast/fwd.hpp]: Just out of curiosity: why are these needed?

File: libvast/src/system/accountant.cpp [hunk 7]
```diff
@@ -41,31 +40,10 @@ namespace {
...
-    auto e = make_error(ec::filesystem_error, "failed to open file:", filename);
-    self->quit(e);
-    return;
-  }
-  file << "timestamp\tnode\taid\tactor_name\tkey\tvalue\n";
-  if (!file)
-    self->quit(make_error(ec::filesystem_error));
-  VAST_DEBUG(self, "kicks off flush loop");
-  self->send(self, flush_atom::value);
 #if VAST_LOG_LEVEL >= CAF_LOG_LEVEL_INFO
```
Comment (mavam) [inline libvast/src/system/accountant.cpp]: This is different from your changes, but I find this pre-processor conditional not very intuitive. To me, it reads like "only perform this when logging is enabled". Can we have a clearer macro for that?

File: integration/default_set.yaml [hunk 8]
```diff
@@ -41,7 +41,7 @@ tests:
       - command: export -N ascii 'orig_bytes > 1k && orig_bytes < 1Ki'
       - command: export -N ascii ':string == "OrfTtuI5G4e" || :port == 67/udp'
       - command: export -N ascii '#type == "zeek.conn" && resp_h == 192.168.1.104'
-      - command: export -N ascii '#type != "zeek.conn"'
+      - command: export -N ascii '#type != "zeek.conn" && \#type != "vast.account"'
```
Comment (mavam) [inline integration/default_set.yaml]: I find this more idiomatic: `#type !in {"zeek.conn", "vast.account"}`.

File: integration/integration.py [hunk 9]
```diff
@@ -173,34 +186,72 @@ def try_wait(process, timeout):
             LOGGER.debug('standard error:')
             for line in open(stderr).readlines()[-100:]:
                 LOGGER.debug(f'    {line}')
+            return result
         # Perform baseline update or comparison.
         baseline = baseline_dir / f'{step_id}.ref'
+        output_format = deduce_format(step.command)
```
Comment (mavam) [inline integration/integration.py]: Couldn't you just try to attempt to parse the output as JSON with builtin Python tools?

File: libvast/src/system/accountant.cpp (hunk 2 above)
Comment (tobim) [reply libvast/src/system/accountant.cpp]: This was added in an attempt do debug a problem. I forgot to take it out. I'm thinking of creating a new macro for ad-hoc debugging. One that would be detected by a pre-commit hook and reject the commit. Maybe `clang-tidy` can help with that.

File: integration/default_set.yaml (hunk 8 above)
Comment (tobim) [reply integration/default_set.yaml]: This does currently not work.

File: integration/integration.py (hunk 9 above)
Comment (tobim) [reply integration/integration.py]: What would be the advantage of that approach?

File: libvast/vast/fwd.hpp (hunk 6 above)
Comment (tobim) [reply libvast/vast/fwd.hpp]: Are you asking why these functions exist in the first place or why are they forward declared here?

File: libvast/vast/system/accountant.hpp [hunk 10]
```diff
@@ -82,6 +84,12 @@ struct accountant_state {
 
   accountant_state(accountant_type::stateful_base<accountant_state>* self);
   void command_line_heartbeat();
+
+  size_t slice_size;
+  table_slice_builder_ptr builder;
+  using downstream_manager = caf::broadcast_downstream_manager<table_slice_ptr>;
```
[the remark this answers is not in the document]
Comment (tobim) [reply libvast/vast/system/accountant.hpp]: `accountant_state` deserves a cleanup. It will be pushed shortly.

File: libvast/src/system/accountant.cpp (hunk 5 above)
Comment (tobim) [reply libvast/src/system/accountant.cpp]: This was actually possible if `finish_slice()` was called from the exit handler. Rectified.

File: libvast/src/system/accountant.cpp (hunk 4 above)
Comment (tobim) [reply libvast/src/system/accountant.cpp]: Fixed

File: libvast/src/system/accountant.cpp (hunk 1 above)
Comment (tobim) [reply libvast/src/system/accountant.cpp]: `clang-format` keeps this entire behavior as it is...

File: integration/integration.py (hunk 9 above)
Comment (mavam) [reply integration/integration.py]: It seemed to be simpler and more maintainable than looking at the command structure. This was just an idea, though, no need to budge.

File: libvast/src/system/accountant.cpp (hunk 1 above)
Comment (mavam) [reply libvast/src/system/accountant.cpp]: Okay.

File: libvast/vast/fwd.hpp (hunk 6 above)
Comment (mavam) [reply libvast/vast/fwd.hpp]: The former.

**Comment by mavam** [comment]
Dang, I noticed one more thing that was missing right after merging: could you add a changelog entry that describes how one can access the accountant data?

File: libvast/src/system/accountant.cpp (hunk 7 above)
Comment (tobim) [reply libvast/src/system/accountant.cpp]: > To me, it reads like "only perform this when logging is enabled".

That is correct however. We don't want to print the heartbeat if verbosity is set to `WARNING` or lower.

Comment (mavam) [reply libvast/src/system/accountant.cpp]: Okay, I think I understand the technical aspect. But wouldn't you agree that this is a hard-to-read of conveying "logging enabled?"
